Add support of Ascend NPU devices

shakhat opened the pull request:
This patch adds support of Ascend NPU devices, current it works only in inference mode and only for PyTorch models. 
For environments with multiple NPU cards it is possible to specify card index by providing it after ":", e.g. "npu:5"

Example of detection in a video file:
```
(.venv) ~/work/ultralytics$ yolo predict detect model=yolov8n.pt source=Rush-Hour-100.mp4 verbose=True device=npu:5                                                       
Ultralytics YOLOv8.2.26 🚀 Python-3.8.10 torch-2.2.0 Ascend NPU:5 (Ascend910ProB, 32768MiB)                  
YOLOv8n summary (fused): 168 layers, 3151904 parameters, 0 gradients, 8.7 GFLOPs                                                                                                     
                                                                                                                                                                                     
[W OpCommand.cpp:88] Warning: [Check][offset] Check input storage_offset[%ld] = 0 failed, result is untrustworthy1 (function operator())
[W VariableFallbackKernel.cpp:51] Warning: CAUTION: The operator 'torchvision::nms' is not currently supported on the NPU backend and will fall back to run on the CPU. This may have
 performance implications. (function npu_cpu_fallback)                                                                                                                               
WARNING ⚠️ NMS time limit 2.050s exceeded                                                                                                                                             
video 1/1 (frame 1/100) /home/ilya/work/ultralytics/Rush-Hour-100.mp4: 384x640 2 persons, 14 cars, 8 motorcycles, 1 bus, 11706.9ms
video 1/1 (frame 2/100) /home/ilya/work/ultralytics/Rush-Hour-100.mp4: 384x640 1 person, 14 cars, 3 motorcycles, 1 bus, 8.7ms
video 1/1 (frame 3/100) /home/ilya/work/ultralytics/Rush-Hour-100.mp4: 384x640 1 person, 17 cars, 3 motorcycles, 8.0ms

```

Due to specifics of Ascend libraries the model is recompiled on each invocation making the inference for the first frame very slow (much worse than CPU), but further frames are processed at the full speed. There are some warnings, but they don't look critical. To compare with CPU (aarch64) a single frame inference time is improved from ~450ms to ~9ms. 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Adds support for Huawei Ascend NPU devices, enabling YOLO models to run on this hardware for faster and more flexible inference. 🚀

### 📊 Key Changes
- Introduced detection and handling for "npu" (Huawei Ascend NPU) devices throughout the codebase.
- Ensured predictions are moved to CPU when using NPU, improving compatibility and speed.
- Updated device selection logic to recognize and initialize NPUs, including requirement checks for necessary libraries.
- Enhanced ONNX Runtime support to use the CANNExecutionProvider for NPU inference.
- Improved dependency management for NPU and ONNX environments.

### 🎯 Purpose & Impact
- Expands hardware compatibility, allowing users to run YOLO models on Huawei Ascend NPUs for potentially faster inference. ⚡
- Makes it easier for organizations with Ascend hardware to leverage Ultralytics models.
- Improves flexibility and future-proofs the codebase for more device types.
- Users with NPUs can expect smoother setup and better performance out-of-the-box. 🖥️✨

shakhat committed:
Add support for Ascend NPU devices (#13227)

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

shakhat commented:
I have read the CLA Document and I sign the CLA

glenn-jocher commented:
@shakhat awesome!! I've been thinking of adding support for Huawei Ascend GPUs but didn't know where to start, so I'm super happy to see this PR.

shakhat committed:
Add support of ONNX models on Ascend NPU devices

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

shakhat commented:
A short summary:
 * The patch adds support for native models (*.pt) and models in ONNX format (max opset 15)
 * Only inference (`yolo predict detect`)
 * Devices: Ascend910ProB and Ascend310
 * Performance is somewhat contradictory - reported inference time is definitely better than CPU, but end-to-end time is worse, hope profiling can reveal the reason

glenn-jocher commented:
@shakhat got it! Is there a way to implement for training as well?

glenn-jocher committed:
Merge branch 'main' into ascend-npu

shakhat commented:
@glenn-jocher so far the training fails with an internal error... I will look if there is a workaround 
```
$ yolo detect train data=coco8.yaml model=yolov8n.yaml epochs=100 imgsz=640 device=npu
...
EZ9999: Inner Error!                                                                                                                                                                 
EZ9999  Kernel task happen error, retCode=0x26, [aicore exception].[FUNC:PreCheckTaskErr][FILE:task_info.cc][LINE:1677]                                                              
        TraceBack (most recent call last):                                                                                                                                           
        The error from device(0), serial number is 2, there is an aicore error, core id is 0, error code = 0x200000010, dump info: pc start: 0x1000124080d98000, current: 0x124080d981b4, vec error info: 0x14effe0f, mte error info: 0x65, ifu error info: 0x1fe3b24f45b80, ccu error info: 0x65, cube error info: 0x45, biu error info: 0, aic error mask: 0x65000200d000288, para base: 0x1240c0459c00, errorStr: Illegal instruction, which is usually caused by unaligned UUB addresses.The write address of the MTE instruction is out of range.[FUNC:PrintCoreErrorInfo][FILE:device_error_proc.cc][LINE:535] 
```

glenn-jocher commented:
@shakhat Thanks for the update! It seems like the NPU training integration is hitting some deep-seated issues. I'll dive into the error logs and see if there's a potential fix or workaround. Will keep you posted! 🛠️

glenn-jocher committed:
Merge branch 'main' into ascend-npu

glenn-jocher committed:
Merge branch 'main' into ascend-npu

glenn-jocher committed:
Merge branch 'main' into ascend-npu

shakhat committed:
Improve postprocessing performance on Ascend NPU devices

Postprocessing operator 'torchvision::nms' is not supported by NPU
and automatically falls back to CPU resulting in high performance
loss (operation takes up to 2 sec.). By placing all inference
results to CPU the postprocessing takes reasonable ~12ms.

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Merge branch 'main' into ascend-npu

UltralyticsAssistant committed:
Merge branch 'main' into ascend-npu

UltralyticsAssistant committed:
Merge branch 'main' into ascend-npu

UltralyticsAssistant committed:
Merge branch 'main' into ascend-npu

UltralyticsAssistant committed:
Merge branch 'main' into ascend-npu

UltralyticsAssistant committed:
Merge branch 'main' into ascend-npu

UltralyticsAssistant committed:
Merge branch 'main' into ascend-npu

UltralyticsAssistant committed:
Merge branch 'main' into ascend-npu

UltralyticsAssistant committed:
Merge branch 'main' into ascend-npu

UltralyticsAssistant committed:
Merge branch 'main' into ascend-npu

UltralyticsAssistant committed:
Merge branch 'main' into ascend-npu

UltralyticsAssistant committed:
Merge branch 'main' into ascend-npu

UltralyticsAssistant committed:
Merge branch 'main' into ascend-npu

UltralyticsAssistant committed:
Merge branch 'main' into ascend-npu

Devcode518 commented:
> @shakhat感谢更新！NPU培训整合似乎触及了一些深层次的问题。我将深入研究错误日志，看看是否有潜在的修复或解决方法。会随时通知你的！🛠️

Hello, may I ask if training is now supported on NPU? For example, the Ascend 910A.

shakhat commented:
> > @shakhat感谢更新！NPU培训整合似乎触及了一些深层次的问题。我将深入研究错误日志，看看是否有潜在的修复或解决方法。会随时通知你的！🛠️
> 
> Hello, may I ask if training is now supported on NPU? For example, the Ascend 910A.

Unfortunately I do not have access to the hardware at this moment :-( But on pypi there are some new torch-npu releases, so hopefully the support may become better.

glenn-jocher commented:
@Devcode518 @shakhat Currently, training on Ascend NPUs is still not officially supported in Ultralytics. The initial PR added inference-only capabilities, and we've seen the same internal errors when attempting training workflows. We're monitoring the new torch-npu releases and hope to revisit this integration when the underlying libraries mature further. If anyone in the community manages to get training working on Ascend hardware, we'd welcome your contributions and insights.
